perception/mil_vision/ros_tools/colorspaces.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python
+from __future__ import division, print_function
+import numpy as np
+import cv2
+import argparse
+
+
+def viz_colorspaces(rgb_image, viz_scale):
+
+    '''
+    Returns image that shows the channels of the BGR, HSV, and LAB colorspace
+    representations of a given image
+    '''
+
+    w = int(rgb_image.shape[1] * viz_scale)
+    h = int(rgb_image.shape[0] * viz_scale)
+    viz = np.zeros((h * 3, w * 4, 3), dtype=np.uint8)
+
+    rgb_small = cv2.resize(rgb_image, dsize=(w, h))
+    hsv = cv2.cvtColor(rgb_small, cv2.COLOR_BGR2HSV)
+    lab = cv2.cvtColor(rgb_small, cv2.COLOR_BGR2LAB)
+
+    viz[h:h * 2, :w] = rgb_small
+    viz[:h, w:w * 2, :], viz[:h, w * 2:w * 3, :], viz[:h, w * 3:w * 4, :] = \
+        map(lambda x: np.repeat(x[:, :, np.newaxis], 3, axis=2), cv2.split(rgb_small))
+    viz[h:h * 2, w:w * 2, :], viz[h:h * 2, w * 2:w * 3, :], viz[h:h * 2, w * 3:w * 4, :] = \
+        map(lambda x: np.repeat(x[:, :, np.newaxis], 3, axis=2), cv2.split(hsv))
+    viz[h * 2:, w:w * 2, :], viz[h * 2:, w * 2:w * 3, :], viz[h * 2:, w * 3:w * 4, :] = \
+        map(lambda x: np.repeat(x[:, :, np.newaxis], 3, axis=2), cv2.split(lab))
```

Comment:
Compact yes, but pretty messy. I'd think something like this would be a better approach (I haven't tested this entire thing):
```python
# Add label to each channel in each image
rgb = np.copy(rgb_small)
cv2.putText(rgb[:, :, 0], 'R', ...)
cv2.putText(rgb[:, :, 1], 'G', ...)
cv2.putText(rgb[:, :, 2], 'B', ...)
cv2.putText(hsv[:, :, 0], 'H', ...)
# etc for all channels

# Construct 3x3 grid of color spaces where each row is RGB, HSV, or LAB respectively
extract_depth = lambda img: img[:, :, 0], img[:, :, 1], img[:, :, 2]
grid = np.vstack(( np.hstack(extract_depth(rgb)), np.hstack(extract_depth(hsv)), np.hstack(extract_depth(lab)) ))

# The grid needs to be 3 channels deep for displaying
grid = np.repeat(grid[:, :, np.newaxis], 3, axis=2)

# build original RGB small image on the side
blank = np.zeros_like(rgb_small)
side_display = np.vstack((blank, rgb_small, blank))

viz = np.hstack((side_display, grid))
```
That way, no weird formula for any positional setting for the text (it's in the space place on each image) and much more readable imo.

And this could be simplified even more I bet.